### PR TITLE
Tidying up for hierarchical naming schemes

### PR DESCRIPTION
**Pull request by ocaisa**
This removes cross-toolchain dependencies due to problems with GMP compilation with intel compilers. It looks like there is some subtle dependency on the binutils underneath GCC when including the C++ libraries (to be more specific: GMP/6.0.0a builds successully on some OSes with intel/2015a but on all architectures with intel/2015b where any dependency on the system binutils has been removed).

The work with 2015a can be found in #2003

### CODE REVIEW

**Comment by boegel** [comment]
@ocaisa: did you really mean to issue a PR with 45 easyconfigs? :-)

File: easybuild/easyconfigs/b/binutils/binutils-2.25-GCC-4.9.3-binutils-2.25.eb [hunk 1]
```diff
@@ -6,7 +6,7 @@ version = '2.25'
 homepage = 'http://directory.fsf.org/project/binutils/'
 description = "binutils: GNU binary utilities"
 
-toolchain = {'name': 'GCC', 'version': '4.9.3-binutils-%(version)s'}
+toolchain = {'name': 'GCC', 'version': '4.9.3-binutils-%s' % version}
```
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.25-GCC-4.9.3-binutils-2.25.eb]: this is a workaround for https://github.com/hpcugent/easybuild-framework/issues/1402, right?

it should be fixed in the framework, no like this

**Comment by boegel** [comment]
Jenkins: ok to test

File: easybuild/easyconfigs/b/binutils/binutils-2.25-GCC-4.9.3-binutils-2.25.eb (hunk 1 above)
Comment (ocaisa) [inline easybuild/easyconfigs/b/binutils/binutils-2.25-GCC-4.9.3-binutils-2.25.eb]: I need to test with my naming scheme. I'm happy to remove it before the merge.

**Comment by ocaisa** [comment]
Yeah sorry for the 46 easyconfigs but that's the dependency list for CGAL and Mesa

**Comment by boegel** [comment]
@ocaisa: 45 is a lot, and can be a pain to get this merged/tested, but let's see how it goes

I can easily compare these easyconfigs with existing ones via `eb --review-pr`, so in terms of reviewing it's not too bad.

File: easybuild/easyconfigs/b/binutils/binutils-2.25-GCC-4.9.3-binutils-2.25.eb (hunk 1 above)
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.25-GCC-4.9.3-binutils-2.25.eb]: Unless we can't fix https://github.com/hpcugent/easybuild-framework/issues/1402, this should be restored back to using `%(version)s`, indeed.

File: easybuild/easyconfigs/b/binutils/binutils-2.25-intel-2015b.eb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+source_urls = [GNU_SOURCE]
+
+toolchain = {'name': 'intel', 'version': '2015b'}
+
+dependencies = [
+    ('zlib', '1.2.8'),
+]
+
+# make sure that system libraries are also considered by ld
+configopts = '--with-sysroot=/'
```
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.25-intel-2015b.eb]: no gold?

```
configopts = '--with-sysroot=/ --enable-gold --enable-ld=default --enable-plugins'
```

if you change this, you need to update the sanity check bit too

File: easybuild/easyconfigs/b/binutils/binutils-2.25-intel-2015b.eb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+homepage = 'http://directory.fsf.org/project/binutils/'
+description = "binutils-2.22: GNU binary utilities"
+
+sources = [SOURCE_TAR_GZ]
+source_urls = [GNU_SOURCE]
+
+toolchain = {'name': 'intel', 'version': '2015b'}
+
+dependencies = [
+    ('zlib', '1.2.8'),
```
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.25-intel-2015b.eb]: flex/Bison are also (build)deps in other binutils easyconfigs?

File: easybuild/easyconfigs/b/binutils/binutils-2.25-intel-2015b.eb [hunk 4]
```diff
@@ -0,0 +1,32 @@
+easyblock = 'ConfigureMake'
+
+name = 'binutils'
+version = '2.25'
+
+homepage = 'http://directory.fsf.org/project/binutils/'
+description = "binutils-2.22: GNU binary utilities"
```
Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.25-intel-2015b.eb]: don't mention the (wrong ;)) version in the description

File: easybuild/easyconfigs/q/Qt/Qt-4.8.6-intel-2015b-GLib-2.44.1.eb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+name = 'Qt'
+version = '4.8.6'
+
+homepage = 'http://qt-project.org/'
```
Comment (boegel) [inline easybuild/easyconfigs/q/Qt/Qt-4.8.6-intel-2015b-GLib-2.44.1.eb]: recent Qt easyconfigs use `qt.io`

some other (style) changes are also missing in this easyconfig

File: easybuild/easyconfigs/g/GMP/GMP-6.0.0a-intel-2015b.eb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+homepage = 'http://gmplib.org/'
+description = """GMP is a free library for arbitrary precision arithmetic, 
+operating on signed integers, rational numbers, and floating point numbers. """
+
+toolchain = {'name': 'intel', 'version': '2015b'}
+toolchainopts = {'optarch': True, 'pic': True, 'precise': True}
+
+sources = [SOURCELOWER_TAR_BZ2]
+source_urls = ['http://ftp.gnu.org/gnu/gmp']
```
Comment (boegel) [inline easybuild/easyconfigs/g/GMP/GMP-6.0.0a-intel-2015b.eb]: `GNU_SOURCE`

File: easybuild/easyconfigs/g/GMP/GMP-6.0.0a-intel-2015b.eb [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+name = 'GMP'
+version = '6.0.0a'
+
+homepage = 'http://gmplib.org/'
+description = """GMP is a free library for arbitrary precision arithmetic, 
+operating on signed integers, rational numbers, and floating point numbers. """
+
+toolchain = {'name': 'intel', 'version': '2015b'}
+toolchainopts = {'optarch': True, 'pic': True, 'precise': True}
```
Comment (boegel) [inline easybuild/easyconfigs/g/GMP/GMP-6.0.0a-intel-2015b.eb]: optarch is enabled by default, so remove?

**Comment by ocaisa** [comment]
`binutils` seems to be problematic with intel compilers once you include ld.gold

**Comment by ocaisa** [comment]
Test report by @ocaisa
**FAILED**
Build succeeded for 45 out of 47 (45 easyconfigs in this PR)
Linux centos linux 7.1.1503, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.9
See https://gist.github.com/8c061c971b42b87b5190 for a full test report.

**Comment by ocaisa** [comment]
Test report by @ocaisa
**SUCCESS**
Build succeeded for 1 out of 1 (45 easyconfigs in this PR)
Linux centos linux 7.1.1503, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.9
See https://gist.github.com/98dc77f68f642281b81a for a full test report.

**Comment by ocaisa** [comment]
Test report by @ocaisa
**SUCCESS**
Build succeeded for 45 out of 45 (45 easyconfigs in this PR)
Linux centos linux 7.1.1503, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.9
See https://gist.github.com/eef111eb2e0b064940f1 for a full test report.

**Comment by ocaisa** [comment]
Ok, the only issue remaining (I think), is whether to stay with a binutils in the toolchain (without ld.gold) or specifically use the binutils built with the underlying GCC. The latter is a bit problematic for me, the underlying GCC is not directly part of the intel toolchain, so would not get picked up by a subtoolchain search.

**Comment by ocaisa** [comment]
Actually this will never work with TCL modules will it? Since binutils is already loaded with GNU, we just end up in a conflict. Maybe I just revert to the underlying binutils?

**Comment by boegel** [comment]
@ocaisa: please sync this with `develop`?

If there's a conflict at some point, Jenkins should spot/report it.

Where would there be a conflict exactly?

**Comment by ocaisa** [comment]
Shouldn't the fact that GNU has a different binutils dependency lead to a conflict? In lmod it triggers a swap not an error, in TCL doesn't it trip an error?

On 7 Oct 2015 18:20, "Kenneth Hoste" <notifications@github.com<mailto:notifications@github.com>> wrote:

@ocaisa<https://github.com/ocaisa>: please sync this with develop?

If there's a conflict at some point, Jenkins should spot/report it.

Where would there be a conflict exactly?

—
Reply to this email directly or view it on GitHub<https://github.com/hpcugent/easybuild-easyconfigs/pull/2013#issuecomment-146250297>.


------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------
Forschungszentrum Juelich GmbH
52425 Juelich
Sitz der Gesellschaft: Juelich
Eingetragen im Handelsregister des Amtsgerichts Dueren Nr. HR B 3498
Vorsitzender des Aufsichtsrats: MinDir Dr. Karl Eugen Huthmacher
Geschaeftsfuehrung: Prof. Dr.-Ing. Wolfgang Marquardt (Vorsitzender),
Karsten Beneke (stellv. Vorsitzender), Prof. Dr.-Ing. Harald Bolt,
Prof. Dr. Sebastian M. Schmidt
------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------

**Comment by boegel** [comment]
@ocaisa: It's still not clear to me where the conflict would be... Which easyconfig should result in having two binutils versions loaded at the same time?

Are you taking into account that build dependencies are not loaded 'at runtime', i.e. when the generated module gets loaded?

**Comment by ocaisa** [comment]
Test report by @ocaisa
**SUCCESS**
Build succeeded for 38 out of 38 (27 easyconfigs in this PR)
Linux centos linux 7.1.1503, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.9
See https://gist.github.com/53dc48245e838d228f29 for a full test report.

**Comment by boegel** [comment]
@ocaisa: https://github.com/ocaisa/easybuild-easyconfigs/pull/12 resolve the conflicts caused by #2028 and #2043

**Comment by ocaisa** [comment]
Ok, this PR no longer makes any sense, I accept defeat! I'll make a PR for GMP and leave it at that.

**Comment by boegel** [comment]
@ocaisa: why doesn't it make sense anymore?

**Comment by ocaisa** [comment]
The original huge eb list was all based around the dependecies new CGAL and Mesa builds with `intel/2015b`, there are already now builds of both in the develop branch so it doesn't really make much sense any more. The only thing really left is GMP

**Comment by boegel** [comment]
@ocaisa: well OK, can you rework this PR to kick out what you think is no longer useful, and keep around what you think is useful?

**Comment by ocaisa** [comment]
Cross-toolchains deps can create problems in a hierarchical MNS - if GNU sits low in a toolchain hierarchy but were reloaded in a HMNS, any GNU built apps common to higher-up components of the hierarchy suddenly take priority...I might only want Autotools but I get swaps for everything that comes with extending the hierarchy

File: easybuild/easyconfigs/b/binutils/binutils-2.25-GCC-4.9.3-binutils-2.25.eb (hunk 1 above)
Comment (ocaisa) [inline easybuild/easyconfigs/b/binutils/binutils-2.25-GCC-4.9.3-binutils-2.25.eb]: Actually I think this should stay, this works regardless of MNS, the other only works with the flat scheme, why leave something broken inside while we wait for a framework  fix?

Comment (boegel) [inline easybuild/easyconfigs/b/binutils/binutils-2.25-GCC-4.9.3-binutils-2.25.eb]: Yeah, it's OK to merge this in until we fix https://github.com/hpcugent/easybuild-framework/issues/1402.

Style shouldn't win over functionality.

**Comment by boegel** [comment]
@ocaisa: ping?

**Comment by ocaisa** [comment]
This is completely outdated now (especially with `--include-minimal-toolchains` and `intel/2016a` coming) anyway so it's dead...closing!

**Pull request closed without merging**